highfidelity/hifi
# Unity Avatar Exporter

This PR is not a release blocker because it does not affect Interface builds and is not a complete feature yet.

Addresses: https://highfidelity.manuscript.com/f/cases/edit/19924

AvatarExporter.cs - the exporter script itself
avatarExporter.unitypackage - the package containing AvatarExporter.cs which can be added to a Unity Project
packager.bat - generates the avatarExporter.unitypackage with AvatarExporter.cs.

Dependent on PR https://github.com/highfidelity/hifi/pull/14486 to use the exported fst in Interface, but does not need to be merged after that PR as it is independent of hifi builds.

Pre-Merge and Post-Merge Test Plan:
- Download https://github.com/highfidelity/hifi/blob/e5792aa85f50be800d01973bfbf565afcf5397c6/tools/unity-avatar-exporter/avatarExporter.unitypackage
and
https://hifi-content.s3.amazonaws.com/davidback/avatars/artemis_test.fbx
- With Unity installed, open a new Unity project
- Open the above avatarExporter.unitypackage and import it into your Unity project and verify a High Fidelity menu appears
- Either select Assets -> Import New Asset and choose the artemis_test.fbx you saved above, or drag the artemis_test.fbx into the Assets folder at the bottom of Unity
- Verify with no asset selected that both Export New Avatar and Update Avatar menu options under High Fidelity are greyed out
- Select a different besides artemis_test (such as SampleScene in Scenes) and verify when choosing Export New Avatar that a dialog is displayed saying to please select an fbx avatar
- Select artemis_test and verify when choosing Export New Avatar that a dialog is displaying saying to please set Animation Type to Humanoid
- Using the Animation Type dropdown in the Inspector on the right, set artemis_test Animation Type to Humanoid and hit Apply
- Hit Configure and delete the Hips bone and verify when exporting the avatar that an error is displaying saying a there is no Hips bone
- Undo the last bone deletion, and delete the Spine bone and verify when exporting the avatar that an error is displaying saying a there is no Spine bone
- Undo the last bone deletion, and delete the Chest bone and verify when exporting the avatar that an error is displaying saying a there is no Chest bone
- Undo the last bone deletion, and delete the UpperChest bone, and export the avatar
- Verify a folder directory is opened that defaults to user's Documents folder, choose Select Folder, and verify a folder named artemis_test is created containing artemis_test.fbx and artemis_test.fst
- Open artemis_test.fst and verify there is a line "jointMap = Spine2 = XXX" where XXX is the bone set to Chest when hitting Configure under Animation Type, and it will not have any "jointMap = Spine1 = YYY" defined
- Undo the last bone deletion
- Verify the Update Avatar menu option is now enabled but only when artemis_test asset is selected
- Choose Update Avatar and verify the artemis_test.fst is updated to have "jointMap = Spine2 = XXX" (where XXX is the bone set to UpperChest in Configure) and "jointMap = Spine1 = YYY" (where YYY is the bone set to Chest in Configure)
- Upload both the artemis_test.fst and artemis_test.fbx to an S3 repository (I can upload them for you if you need)
- Copy the fst URL, and open Interface with PR https://github.com/highfidelity/hifi/pull/14486, and set that as your avatar
- Verify an untextured Artemis avatar is applied and looks and animates as expected
